ipld/src/selector/mod.rs

```diff
@@ -236,9 +235,126 @@ impl Selector {
-    pub fn explore(self, _ipld: &Ipld, _p: &PathSegment) -> Option<Selector> {
-        // TODO
-        todo!()
+    pub fn explore(self, ipld: &Ipld, p: &PathSegment) -> Option<Selector> {
+        use Selector::*;
+        match self {
+            ExploreAll { next } => Some(*next),
+            ExploreFields { mut fields } => match ipld {
+                Ipld::Map(m) => match p {
+                    // Check if field exists on ipld, then explore selector
+                    PathSegment::String(s) => {
+                        m.get(s)?;
+                        fields.remove(s)
+                    }
+                    PathSegment::Int(i) => {
+                        let key = format!("{}", i);
+                        m.get(&key)?;
+                        fields.remove(&key)
+                    }
+                },
+                // Using ExploreFields for list is supported feature in go impl
+                Ipld::List(l) => {
+                    // Check to make sure index is within bounds
+                    if p.to_index()? >= l.len() {
+                        return None;
+                    }
+                    match p {
+                        PathSegment::String(s) => fields.remove(s),
+                        PathSegment::Int(i) => fields.remove(&format!("{}", i)),
+                    }
+                }
+                _ => None,
+            },
+            ExploreIndex { index, next } => match ipld {
+                Ipld::List(l) => {
+                    let i = p.to_index()?;
+                    if i != index || i >= l.len() {
+                        None
+                    } else {
+                        // Path segment matches selector index
+                        Some(*next)
+                    }
+                }
+                _ => None,
+            },
+            ExploreRange { start, end, next } => {
+                match ipld {
+                    Ipld::List(l) => {
+                        let i = p.to_index()?;
+                        // Check to make sure index is within list bounds
+                        if i < start || i >= end || i >= l.len() {
+                            None
+                        } else {
+                            // Path segment is within the selector range
+                            Some(*next)
+                        }
+                    }
+                    _ => None,
+                }
+            }
+            ExploreRecursive {
+                current,
+                sequence,
+                limit,
+                stop_at,
+            } => {
+                let next = match current {
+                    Some(s) => s.explore(ipld, p)?,
+                    None => sequence.clone().explore(ipld, p)?,
+                };
+
+                if !has_recursive_edge(&next) {
+                    return Some(ExploreRecursive {
+                        sequence,
+                        current: Some(next.into()),
```

Comment:
```suggestion
                        current: Some(Box::new(next)),
```
Somewhat subjective, but `From` and `Into` are super generic and it's not obvious which conversion is taking place here without carefully reading the code

Reply:
yeah I guess the counter point to that is that it's harder to read the logic of what's happening with the `Box::new(..)` clutter when that detail (the fact that it's boxed) doesn't matter. I'm happy to change it if you guys think this is more readable

Reply:
Yeah that's a very valid point, `into` somewhat hides the stuff that doesn't matter here. But it unfortunately can do all kinds of conversions, which makes this (though easier to read) slightly harder to understand :)

A third option is always to add a `boxed()` method to every type, might slightly get rid of the `Box::new` clutter without hiding the kind of conversion. But then the reader does need to know that it does the same as `Box::new`, so it's not perfect either.